plugins/intrinsics/src/main/java/org/qbicc/plugin/intrinsics/core/CoreIntrinsics.java

```diff
@@ -285,15 +289,92 @@ public static void registerJavaLangThreadIntrinsics(CompilationContext ctxt) {
+        ClassTypeDescriptor vmHelpersDesc = ClassTypeDescriptor.synthesize(classContext, "org/qbicc/runtime/main/VMHelpers");
+        ClassTypeDescriptor pthreadPtrDesc = ClassTypeDescriptor.synthesize(classContext, "org/qbicc/runtime/posix/PThread$pthread_t_ptr");
+        ClassTypeDescriptor voidPtrDesc = ClassTypeDescriptor.synthesize(classContext, "org/qbicc/runtime/CNative$void_ptr");
+        ClassTypeDescriptor jimVMDesc = ClassTypeDescriptor.synthesize(classContext, "jdk/internal/misc/VM");
+
         MethodDescriptor returnJlt = MethodDescriptor.synthesize(classContext, jltDesc, List.of());
+        MethodDescriptor voidDesc = MethodDescriptor.synthesize(classContext, BaseTypeDescriptor.V, List.of());
+        MethodDescriptor voidIntDesc = MethodDescriptor.synthesize(classContext, BaseTypeDescriptor.V, List.of(BaseTypeDescriptor.I));
+        MethodDescriptor booleanDesc = MethodDescriptor.synthesize(classContext, BaseTypeDescriptor.Z, List.of());
 
-        StaticIntrinsic currentThread = (builder, target, arguments) -> builder.currentThread();
+        StaticIntrinsic nopStatic = (builder, target, arguments) -> ctxt.getLiteralFactory().zeroInitializerLiteralOfType(target.getType().getReturnType());
+        InstanceIntrinsic nopInstance = (builder, instance, target, arguments) -> ctxt.getLiteralFactory().zeroInitializerLiteralOfType(target.getType().getReturnType());
 
+        /* registerNatives */
+        intrinsics.registerIntrinsic(jltDesc, "registerNatives", voidDesc, nopStatic);
+
+        /* public static native Thread currentThread(); */
+        StaticIntrinsic currentThread = (builder, target, arguments) -> builder.currentThread();
         intrinsics.registerIntrinsic(jltDesc, "currentThread", returnJlt, currentThread);
 
-        StaticIntrinsic nop = (builder, target, arguments) -> ctxt.getLiteralFactory().zeroInitializerLiteralOfType(target.getType().getReturnType());
+        /* special function that will be passed to pthread_create */
+        MethodDescriptor threadWrapperDesc = MethodDescriptor.synthesize(classContext, voidPtrDesc, List.of(voidPtrDesc));
+        StaticIntrinsic threadWrapper = (builder, target, arguments) -> {
+            Value thread = arguments.get(0);
+
+            /* set current thread */
+            ValueHandle qbiccCurrentThreadPointer = builder.pointerHandle(ctxt.getCurrentThreadLocalSymbolLiteral());
+            builder.store(qbiccCurrentThreadPointer, thread, MemoryAtomicityMode.NONE);
+
+            /* call "run" method of thread object */
+            VirtualMethodElementHandle runHandle = (VirtualMethodElementHandle)builder.virtualMethodOf(thread, jltDesc, "run", voidDesc);
+            builder.call(runHandle, List.of());
+
+            /* set java.lang.Thread.threadStatus to terminated */
+            Value terminatedState = builder.load(builder.staticField(jimVMDesc, "JVMTI_THREAD_STATE_TERMINATED", BaseTypeDescriptor.I), MemoryAtomicityMode.NONE);
```

Comment:
I think we can also hardcode `JVMTI_THREAD_STATE_TERMINATED` as the value is from the jvmti spec